Add WIP OpenGL Pixel History code for PostMod pixel value

orsonbraines opened the pull request:
<!--
Before submitting a pull request you are strongly recommended to read the
docs/CONTRIBUTING.md file which gives some information on how to prepare a
change:

https://github.com/baldurk/renderdoc/blob/v1.x/docs/CONTRIBUTING.md

For small changes you don't have to read the document end to end, but should at
least look at the sections on how to ensure your code and commits are formatted
according to the style requirements.
-->

## Description

<!--
Describe here what your pull request changes and why it should happen. For small
changes which are obvious this can just be a line or two - even the commit
message is sometimes enough.
-->

The OpenGL Pixel History feature is hidden behind the OpenGL_PixelHistory config toggle.
For now, we only run an occlusion query and then replay the events
and read back the pixel values to return to the UI.

orsonbraines committed:
Add WIP OpenGL Pixel History code for PostMod pixel value

The OpenGL Pixel History feature is hidden behind the OpenGL_PixelHistory config toggle.
For now, we only run an occlusion query and then replay the events
and read back the pixel values to return to the UI.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -0,0 +1,225 @@
+    history[i].preMod = history[i - 1].postMod;
+  }
+
+  PixelHistoryDestroyResources(m_pDriver, resources);
```
This destroy will also need to happen above in the early `return history` if no modifying events are found.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -0,0 +1,225 @@
+  for(size_t i = 0; i < modEvents.size(); i++)
+  {
+    driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+    driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_Full);
```
Since this is GL you should restore the framebuffer binding after replaying the draw, since e.g. one of the events below between draws might interact with the current framebuffer.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -0,0 +1,225 @@
+rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                                          const rdcarray<EventUsage> &events, int x, int y)
+{
+  driver->glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
```
If this clear is intended to clear your own framebuffer you should bind it explicitly first, as currently here this seems like it only works by coincidence that the framebuffer was just created (and so implicitly bound because this is GL).

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -0,0 +1,225 @@
+    }
+    else
+    {
+      driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
```
Is there a reason to bind your own framebuffer here, for only the occlusion query?

orsonbraines replied:
I think you're right, we shouldn't need to bind it here

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -0,0 +1,225 @@
+      driver->glDisable(eGL_DEPTH_TEST);
+      driver->glDisable(eGL_STENCIL_TEST);
+      driver->glEnable(eGL_SCISSOR_TEST);
+      driver->glScissor(x, y, 1, 1);
```
There are a couple more state bits that need to be disabled here (e.g. backface culling, sample masking) which could affect pixels being written to.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -0,0 +1,225 @@
+      driver->ReplayLog(events[i].eventId, events[i].eventId, eReplay_OnlyDraw);
+
+      int result;
+      driver->glGetQueryObjectiv(resources.occlusionQuery, eGL_QUERY_RESULT, &result);
```
You must call `glEndQuery` before querying the result here.

Also fetching the result here inside the loop one by one will force the GPU to sync after every draw. It is better to have a query and query result per potential event, and only fetch them after submitting all of the queries to only sync once.

orsonbraines replied:
When I put in glEndQuery I get these error messages shwoing up in the log:

```RDOC 001648: [13:40:06]        gl_driver.cpp(3223) - Log     - 'GL_INVALID_OPERATION error generated. <target> does not have an active query.'```

After doing some debugging, it looks like when calling ReplayLog, gl_driver.cpp:5276 ends the query before it returns. Do you know why that would be?

baldurk replied:
Ah I forgot that GL required strict closing of queries to avoid problems. You can call `WrappedOpenGL::SetFetchCounters` to temporarily disable this when replaying a single draw, so that you can control the counters yourself. See the usage in `gl_counters.cpp`.

baldurk reviewed (changes requested):
I think this is a good start, I've noted feedback where there are some things to improve.

orsonbraines commented:
One thing that we noticed when this is that the colours being displayed for the hover and the Pixel History look different than how they appear in the texture view itself. It looks like the UI is doing an sRGB onversion that is messing the colours up. Is this something we need to try and fix in our code or is a problem with the UI code?

![renderdocColour](https://user-images.githubusercontent.com/13425717/173893047-1468cf4b-8335-45b3-b217-90d659d75e94.png)

orsonbraines committed:
OpenGL Pixel History address PR feedback

- add missing destroy resources call
- restore framebuffer binding after replaying draw
- disable more state bits for occlusion querying
- read back query results after executing all queries

baldurk commented:
Yes it looks like the behaviour was inconsistent at best in the texture viewer swatch, and largely wrong in the pixel history view. Although these are intended just for general at-a-glance it's better to treat them as linear data and display them to be interpreted as sRGB, rather than trying to unconvert. I've pushed a change to improve that behaviour.

orsonbraines committed:
use WrappedOpenGL::SetFetchCounters in occlusion query
